File: LC/python/Largest Rectangle in Histogram.py
class Solution:
    # @param height, a list of integer
    # @return an integer
    def largestRectangleArea(self, height):
        stack = []
        height.append(0)
        maxA = 0
        i = 0
        while i < len(height):
            #print(stack)
            if stack == [] or height[i]>=height[stack[-1]]:
                stack.append(i)
                i += 1
            else:
                tmp = stack.pop()
                if stack == []:
                    area = height[tmp] * i
                else:
                    print(i,stack[-1])
                    area = height[tmp] * (i-stack[-1]-1)
                maxA = max(maxA,area)
        return maxA
a = Solution()
b = a.largestRectangleArea([2,1,5,6,2,3])

File: LC/java/Largest Rectangle in Histogram.java
public class Solution {
    public int largestRectangleArea(int[] height) {
        Stack<Integer> s = new Stack<Integer>();
        int i = 0, tmp, area = 0, maxA = 0,com;//com is used to make sure that all items in stack will pop up finally
        
        while(i<height.length+1){
            if(i<height.length){
                com = height[i];
            }
            else{
                com = 0;
            }
            if (s.isEmpty() || height[s.peek()]<com){
                s.push(i);
                i++;
            }
            else{
                tmp = s.pop();
                if (s.isEmpty()){
                    area = height[tmp]*i;
                }
                else{
                    area = height[tmp]*(i-s.peek()-1);
                }
                maxA = Math.max(maxA,area);
            }
            
        }
        return maxA;
    }
}